Implements Non-Blocking Model Hook Support, closes #232

hhff opened the pull request.

hhff commented:
@snewcomer - took a pass at allowing this to work with non-blocking model hooks.

Haven't had a pass at tests yet cause I wanted to get your feedback here?

hhff commented:
Oh BTW - this is working great in an app I'm developing against!

snewcomer commented on addon/components/infinity-loader.js:
```diff
@@ -28,6 +42,10 @@ const InfinityLoaderComponent = Ember.Component.extend({
   },
 
+  _isInfinityPromiseArray: Ember.computed('infinityModel', function() {
+    return (this.get('infinityModel') instanceof InfinityPromiseArray);
```
Naive question - In which case is the infinityModel not an instance of InfinityPromiseArray?

hhff replied:
Good Q - answered below!

snewcomer commented on addon/lib/infinity-promise-array.js:
```diff
@@ -0,0 +1,2 @@
+import Ember from 'ember';
+export default Ember.ArrayProxy.extend(Ember.PromiseProxyMixin);
```
Wondering what the choice of `ArrayProxy` was for?

hhff replied:
I guess it could be ObjectProxy actually...?  Is that what you mean?

snewcomer replied:
yeah that could be one option.  Only b/c it looks like you are concerned with the `content` in the `willInsertElement` method.

hhff replied:
The main reason here for the wrapper however is due to needing "something to mixin `Ember.PromiseProxyMixin`.  Without it, the bound view doesn't update when the promise resolves, meaning you never actually see the InfinityModel's content.

However - I did think about mixing `Ember.PromiseProxyMixin` directly into our `InfinityModel` (which is already an ArrayProxy).

Do you think this would make more sense?  The reason I didn't is because I think it may mean a fundamental change to how we're thinking about the InfinityModel...

hhff replied:
IE - right now our route still handles all things to do with data fetching.  Our InfinityModel only handles building params and storing content.

If we mixed `Ember.PromiseProxyMixin` into it - we'd need to store the initial promise on the InfinityModel (not a bad idea) but at that point we need to ask: "should the InfinityModel have the store injected, and be in charge of actually "making requests"?

at that point we may not even need the route mixin...

snewcomer replied:
@hhff Hm.  At first glance, I like the idea of having the separation of concerns and responsibilities b/w the data and fetching that data.  

What about the `_loadingMore` variable?  Could that be used as the boolean that renders/doesn't render the `infinity-loader` component?

hhff replied:
I agree re: seperation of concerns.  Lets keep the mixin <> model pattern for now!

`_loadingMore` won't help us here unfortunately - as the issue is that the initial change in  `infinityModel.content` (not InfinityPromise) after it's rendered doesn't trigger a template re-render.  When wrapped in a PromiseProxy it's fine (PromiseProxy is the same as a DS.RecordsArray in practice)

I tried a bunch of other stuff like manually triggering a `propertyDidChange` on `[]` and `content` etc after the `pushObjects` call, but nothing worked.  PromiseProxy is the only thing that did.

hhff replied:
If it works w/ ObjectProxy though, we should do that...!

snewcomer replied:
Ah that makes sense.  The template needs a new reference to an array or through updating the content on the proxy for it to re-render.  I'm digging this implementation the more I look at it.

snewcomer replied:
I wonder if the `_loadingMore` variable can be swapped out with `isPending` property on the infinityModel now?

hhff replied:
not all infinityModel instances have `isPending` (in the case I explained about "stashing" infinityModels).  It's only avail on the "first load" directly from a model hook

snewcomer commented on app/templates/components/infinity-loader.hbs:
```diff
@@ -1,7 +1,7 @@
 {{#if hasBlock}}
-  {{yield}}
+  {{yield infinityModelContent}}
```
I'm sure this is already on your radar, but some schweet documentation improvements would be awesome!

hhff replied:
100% - wanted to get your thoughts first!  Next I'll hit tests, then docs

hhff commented:
@snewcomer - technically the infinityModel (passed into the infinity-loader) is *always* a promise array.  The reason I made it continue to accept a "raw" infinity model is for edge cases where users are doing something weird like: 

`loading an infinity model outside a model hook and yielding the result from .then ->` 
`stashing it in a service ->` 
`passing it into a template from the service sometime after the first render`

I've done weird stuff like that before - I can see it being a highly confusing thing for folks who are in the wavy world...

snewcomer commented on addon/components/infinity-loader.js:
```diff
@@ -77,22 +95,29 @@ const InfinityLoaderComponent = Ember.Component.extend({
    */
   _shouldLoadMore() {
-    if (this.get('developmentMode') || typeof FastBoot !== 'undefined' || this.isDestroying || this.isDestroyed) {
+    const consideredReady = !this.get('_isInfinityPromiseArray') || this.get('infinityModel.isFulfilled');
```
The first part of this `||` statement - if it is not the PromiseProxy, then it isn't considered ready to load more?

snewcomer replied:
I guess I would have imagined only the `this.get('infinityModel.isFulfilled')` is needed.  But I may be missing something!

hhff replied:
in the case this is a "raw" infinity model (ie - not wrapped in InfinityPromiseArray) we consider it ready to go (that means the first set of content is resolved). this is the `!this.get('_isInfinityPromiseArray')` case

(also - In that case, `isFulfilled` is undefined - it's only available on PromiseProxies)

however:

in the case it's a PromiseArray, we need to be sure that the first load has actually resolved (so we wait until isFulfilled is true)

snewcomer commented:
@hhff do you think the `route` mixin could be named `data-load` or something like that?  I.e. this mixin seems like it could be used for components as well (one could imagine an activity feed inside of a route or something).

snewcomer commented:
@hhff I'm in favor of this PR!  Just one question above regarding the route mixin name.  Lmk if you want some help with tests/documentation.

hhff commented:
I'd rather we didn't mess with the route mixin, because:
a) it would break a lot of apps currently importing it as it's currently named
b) components should never own or fetch data, as they're lifecycle is not permanent.  it's a pattern i'm not comfortable promoting.  

However - I'd be open to services doing data fetching...

hhff commented:
feel like if we go down that route we might as well make the whole lib service based...

```
Route.extend({
  infinity: service(),
  model() {
    return get(this, 'infinity').infinityModel('posts');
  }
})
```

Thoughts?  This makes sense to me but not sure it makes sense to do for v1.0.0

snewcomer commented:
@hhff Yeah probably not for v1.0.  

And yeah I agree.  I was just thinking about non-route infinite scrolls and how we can make this lib more accessible for those use cases (some nested component that needs infinite scroll).  For example, instead of mixing in the route, one could imagine as a mixin to a controller or service like you mentioned above and having the same facilities to infinite scroll a nested component.  Something we can keep in the back of our minds!

hhff commented:
nice OK @snewcomer !  Got a conference this weekend but will try and sneak some time over the next few days to finish it all off here!

snewcomer commented:
@hhff I think we both have been busy lately :)  lmk if you need any help with this PR!  Happy to help with some docs or testing.

hhff commented:
Thanks @snewcomer !  I should have a chance to finish this PR off on Friday!

hhff commented:
Hi @snewcomer - I took a pass at a few things about a week ago - but have been back to back drilled with work since.

I did learn that the approach above is the best of everything we discussed - main point here is  ArrayProxy works whereas ObjectProxy doesn't for this case (when mixed with the PromiseProxy mixin).

I updated the Ember CLI in hopes it would help with some flaky tests I was seeing - and I added a simple test case to ensure that a non-blocking route will eventually show records.

However - I think in the upgrade the switch to Headless Chrome as the test driver made the calculations in the triggerOffset tests measure incorrectly.

I'm not going to have any time to work on this for at least a month going forward - mid november at least.  Are you down to pickup where I left off (so we can ship 1.0 this year)?

hhff commented:
(i'm only seeing 2 fails locally, FWIW)

snewcomer commented:
@hhff Just got back from vacation!  I'll work to get 1.0 done hopefully in the month of Nov. and help with this PR this coming week.  Awesome job!

hhff commented:
THANKYOU SCOTT!!!  welcome back to git land!!

On Sat, Oct 28, 2017 at 3:20 PM Scott Newcomer <notifications@github.com>
wrote:

> @hhff <https://github.com/hhff> Just got back from vacation! I'll work to
> get 1.0 done hopefully in the month of Nov. and help with this PR this
> coming week. Awesome job!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hhff/ember-infinity/pull/237#issuecomment-340213823>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACu4_P8pCVA2QTRZ9RKAjolRW2sBuR-Lks5sw36WgaJpZM4PSgTF>
> .
>

snewcomer commented:
From a previous PR I did:

Running this test [here](https://github.com/hhff/ember-infinity/blob/master/tests/acceptance/infinity-route-offset-trigger-test.js#L79) and seems as though in test land Ember's event handling has changed and for some reason the height/distance of items in the DOM are also different.  Need to figure out where/why this has changed to be able to modify the tests.

**master ember 2.4 ember-cli-qunit <= 2.1**
<img width="216" alt="screen shot 2017-04-15 at 3 15 58 pm" src="https://cloud.githubusercontent.com/assets/7374640/25067106/7a06d1f0-21ee-11e7-94cb-bb7bcb0b3437.png">

**this branch ember 2.8 ember-cli-qunit > 2.1**
<img width="289" alt="screen shot 2017-04-15 at 3 19 42 pm" src="https://cloud.githubusercontent.com/assets/7374640/25067121/ff4420d4-21ee-11e7-9d11-7f277d7ffc6a.png">

snewcomer commented:
Further testing:

ember-cli-qunit 3.0.1 or 2.2.2 works.  Anything above fails.

snewcomer commented:
@hhff alrighty so it looks like we got a passing test suite!  Had to downgrade to 2.11 and lock ember-cli-qunit to 3.0.1 but all test scenarios passed!  I really like what you did in this PR :+1: 

Let me know what you think or if you want some more work to be done.

hhff commented:
Awesome - amazing work Scott! Will take a look asap!
On Tue, 31 Oct 2017 at 1:41 pm, Scott Newcomer <notifications@github.com>
wrote:

> @hhff <https://github.com/hhff> alrighty so it looks like we got a
> passing test suite! Had to downgrade to 2.11 and lock ember-cli-qunit to
> 3.0.1 but all test scenarios passed! I really like what you did in this PR
> 👍
>
> Let me know what you think or if you want some more work to be done.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/hhff/ember-infinity/pull/237#issuecomment-340843600>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACu4_M0V0vuJJ_GER0nPChBLfqDi1H_xks5sx1urgaJpZM4PSgTF>
> .
>

snewcomer commented:
@hhff just a friendly ping on your PR.  No rush!
